desidlas/training/model.py

```diff
@@ -0,0 +1,304 @@
+    y_nn_classifier = tf.reshape(y_fc4_1, [-1], name='y_nn_classifer')
+    y_fc4_2 = tf.add(tf.matmul(h_fc2_2_drop, W_fc3_2), b_fc3_2)
+    y_nn_offset = tf.reshape(y_fc4_2, [-1], name='y_nn_offset')
+    y_fc4_3 = tf.add(tf.matmul(h_fc2_3_drop, W_fc3_3), b_fc3_3)
+    y_nn_coldensity = tf.reshape(y_fc4_3, [-1], name='y_nn_coldensity')
+
+    # Train and Evaluate the model
+    loss_classifier = tf.add(tf.nn.sigmoid_cross_entropy_with_logits(logits=y_nn_classifier, labels=label_classifier),
+                             l2_regularization_penalty * (tf.nn.l2_loss(W_conv1) + tf.nn.l2_loss(W_conv2) +
+                                                          tf.nn.l2_loss(W_fc1) + tf.nn.l2_loss(W_fc2_1)),
+                             name='loss_classifier')
+    loss_offset_regression = tf.add(tf.reduce_sum(input_tensor=tf.nn.l2_loss(y_nn_offset - label_offset)),
+                                    l2_regularization_penalty * (tf.nn.l2_loss(W_conv1) + tf.nn.l2_loss(W_conv2) +
+                                                                 tf.nn.l2_loss(W_fc1) + tf.nn.l2_loss(W_fc2_2)),
+                                    name='loss_offset_regression')
+    epsilon = 1e-6
+    loss_coldensity_regression = tf.reduce_sum(
+        input_tensor=tf.multiply(tf.square(y_nn_coldensity - label_coldensity),
+                    tf.compat.v1.div(label_coldensity,label_coldensity+epsilon)) +
+        l2_regularization_penalty * (tf.nn.l2_loss(W_conv1) + tf.nn.l2_loss(W_conv2) +
+                                     tf.nn.l2_loss(W_fc1) + tf.nn.l2_loss(W_fc2_1)),
+        name='loss_coldensity_regression')
+
+    optimizer = tf.compat.v1.train.AdamOptimizer(learning_rate=learning_rate)
+
+    cost_all_samples_lossfns_AB = loss_classifier + loss_offset_regression
+    cost_pos_samples_lossfns_ABC = loss_classifier + loss_offset_regression + loss_coldensity_regression
+    # train_step_AB = tf.train.AdamOptimizer(learning_rate=learning_rate).minimize(cost_all_samples_lossfns_AB, global_step=global_step, name='train_step_AB')
+    train_step_ABC = optimizer.minimize(cost_pos_samples_lossfns_ABC, global_step=global_step, name='train_step_ABC')
+    # train_step_C = tf.train.AdamOptimizer(learning_rate=learning_rate).minimize(loss_coldensity_regression, global_step=global_step, name='train_step_C')
+    output_classifier = tf.sigmoid(y_nn_classifier, name='output_classifier')
+    prediction = tf.round(output_classifier, name='prediction')
+    correct_prediction = tf.equal(prediction, label_classifier)
+    accuracy = tf.reduce_mean(input_tensor=tf.cast(correct_prediction, tf.float32), name='accuracy')
+    rmse_offset = tf.sqrt(tf.reduce_mean(input_tensor=tf.square(tf.subtract(y_nn_offset,label_offset))), name='rmse_offset')
+    rmse_coldensity = tf.sqrt(tf.reduce_mean(input_tensor=tf.square(tf.subtract(y_nn_coldensity,label_coldensity))), name='rmse_coldensity')
+
+    variable_summaries(loss_classifier, 'loss_classifier', 'SUMMARY_A')
```

Comment:
What do these lines do?  Please add a comment.

Reply:
done

Reply:
Hi X, I just have another small question. When I reply to your comments, I just reply one by one. This will send you many emails (one reply one email) or just one email that includes all my replies? If it is the first case, I think this may disturb you a lot. So if it is the first case, could you tell me how to do a better replying? If it is the second case, then I will just keep replying like before. Thanks!

Reply:
I like them one by one.  Thx.